designs/0004-closures-fun-types.md

```diff
@@ -0,0 +1,608 @@
+(the rvalue) in `a = b`,
+
+| lvalue type  | legal rvalue types                     |
+|:------------:|:--------------------------------------:|
+| `int(int)`   | `int(int)`, `int(real)`                |
+| `real(real)` | `real(real)`, `int(real)`              |
+| `int(real)`  | `int(real)`                            |
+| `real(int)`  | `real(int)`, `real(real)`, `int(real)` |
+
+
+#### Sized types
+
+The set of sized runtime types is the least set of types such that
+
+* *sized primitive type*: `int` and `real` are sized types,
+* *sized vector type*: `vector[K]` and `row_vector[K]` are sized
+types if `K` is a non-negative integer,
+* *sized matrix type*: `matrix[M, N]` is a sized type if `M` and `N` are
+  non-negative integers,
+* *sized array type*: `T[]` is a sized type if `T` is a sized type, and
+* *sized function type*: `T0(T1,...,TN)` is a sized type if `T0`, `T1`, ..., `TN`
+are sized types.
+
+
+## Lambda Expression Syntax
+
+Syntactically, a lambda expression is represented as a function argument
+list followed by a function body.   For example, in `(real u) { return
+1 / (1 + exp(-u)); }`, the function argument list is `(real u)` and
+the function body is `{ return 1 / (1 + exp(-u)); }`.
+
+As a shorthand, a function argument list followed by an expression is
+taken to be shorthand for the function argument list followed by a
+function body returning that expression.  For example, the lambda
+expression
+
+```
+(real u) { return 1 / (1 + exp(-u)); }
+```
+
+may be replaced with the equivalent
+
+```
+(real u) 1 / 1 + exp(-u)
+```
+
+Lambda expressions behave like other expressions such as `2 + 2`, only
+they denote functions rather than values and have function types.
+
+
+## Changes to Standard Functions
+
+Allow functions to be defined in any scope.  Allow function
+definitions and lambda expressions to
+capture variables in scope by references.  
+
+Deprecate the functions block;  functions declared there can be
+declared in the transformed data block.  
+
+Standard function definitions of the form
+
+```
+T0 f(T1 x1, ..., TN xN) { ... }
+```
+
+may be replaced with
+
+```
+T0(T1, ..., TN) f = (T1 x1, ..., TN xN) { ... }
+```
+
+## Variable declarations
+
+Variable declarations for functions follows the usual syntax, with a
+sized type required for block variables, an unsized type for function
+arguments, and either a sized (as of 2.20) or unsized (future) type
+for a local variable declaration.
+
+
+## Assignment
+
+Nothing changes conceptually about assignment.  The right-hand side
+type must still be assignable to the left-hand side type, which for
+functions, means they have the same function type.  The only question
+is whether to require strict matching of types or support full
+covariance and contravariance.  
+
+## Implementation
+
+Lambdas can be mapped directly to C++ lambdas with default reference
+closures.  This will work because the scope of the compiled C++ is the
+same as that of the Stan program.
```

Comment:
Is it worth thinking about an implementation strategy of compiling away the higher-order functions using something like defunctionalization? That would keep the intermediate representation simpler and wouldn't add more overhead to maintaining the optimizations. On the other hand, it would require more fancy footwork and some thinking.